whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotMirrorImporter.java

```diff
@@ -145,31 +145,59 @@ public void run() {
+        final ExecutorService executor = Executors.newFixedThreadPool(numberOfThreads);
+        final List<TransactionStatus> batchTransactions = processBatchAndTransactions(remainingRecords, processedCount, executor);
+        batchTransactions.forEach(nrtmClientUpdateTransaction::commit);
+    }
+
+    @NotNull
+    private List<TransactionStatus> processBatchAndTransactions(final String[] remainingRecords, final AtomicInteger processedCount, final ExecutorService executor){
+
+        //Transaction annotation does not work with any threaded processing methods
+        final List<TransactionStatus> transactionStatuses = new ArrayList<>();
+        try {
+            Arrays.stream(remainingRecords)
+                    .forEach(record -> {
+                        final TransactionStatus transactionStatus = nrtmClientUpdateTransaction.getTransaction(new DefaultTransactionDefinition()); // create a transaction per record
+                        processRecord(executor, processedCount, record, transactionStatus); // process record
+                        transactionStatuses.add(transactionStatus); // keep the transaction in case there is an issue in another record in the batch
+                    });
+        } catch (Exception e){
+            transactionStatuses.forEach(nrtmClientUpdateTransaction::rollback);
+            throw new IllegalStateException("Unable to persist snapshot", e);
+        } finally {
+            executor.shutdown();
             try {
-                future.get();
-            } catch (Exception e) {
-                throw new RuntimeException("Error processing records", e);
+                executor.awaitTermination(1, TimeUnit.MINUTES);
+            } catch (InterruptedException e) {
+                executor.shutdownNow(); // Interrupted while waiting; force shutdown
+                Thread.currentThread().interrupt();
             }
         }
 
-        executor.shutdown();
+        return transactionStatuses;
+    }
+
+    private void processRecord(final ExecutorService executor, final AtomicInteger processedCount, final String record, final TransactionStatus transactionStatus){
+        try {
+            executor.submit(() -> {
+                try {
+                    final MirrorSnapshotInfo mirrorRpslObject = new ObjectMapper().readValue(record, MirrorSnapshotInfo.class);
+                    final Map.Entry<RpslObject, RpslObjectUpdateInfo> persistedRecord = nrtm4ClientRepository.processSnapshotRecord(mirrorRpslObject);
+                    nrtm4ClientRepository.createIndexes(persistedRecord.getKey(), persistedRecord.getValue());
+                    processedCount.incrementAndGet();
+                } catch (Exception e) {
+                    nrtmClientUpdateTransaction.rollback(transactionStatus);
+                    LOGGER.error("Unable to process record {}", record, e);
+                    throw new IllegalStateException(e);
+                }
+            });
+        } catch (Exception e) {
```

Comment:
I don't think the outside try/catch is useful, unless you expect `submit()` to fail. The record is processed asynchronously. You could return the `Future` and check in the loop whether the job failed, and then stop submitting jobs if any failed.

Reply:
Or use a global `AtomicBoolean` to flag if any job failed and if so, stop all jobs immediately.

Reply:
I go with future for handling errors. So I check for the errors